src/iop/ashift.c

```diff
@@ -2745,9 +2745,9 @@ static int do_get_structure(dt_iop_module_t *module, dt_iop_ashift_params_t *p,
 
   g->fitting = 1;
 
-  dt_pthread_mutex_lock(&g->lock);
+  dt_iop_gui_enter_critical_section(module);
```

Comment:
This here is a read, not a write, of g->buf. As an example of a problem it is fixing: without a lock you might read out a g->buf between free(g->buf) and g->buf = NULL (see below).
As for atomicity question - while some types might be atomic, there's also cache coherency. Different CPU architectures have different memory models, which affect these things. Last and not least are compiler optimisations, how it can re-arrange instructions when there are no memory fences.

Reply:
> you might read out a g->buf between free(g->buf) and g->buf = NULL

That's exactly why I wondered if the lock should not be held _longer_. I've got a feeling I might be incredibly dim again, but I see exactly one piece of code like you describe that is protected by the lock (line 4514) and I don't see how the lock prevents reading the value of g->buf _before_ **both** the free _and_ the = NULL and how that would be any different. Could you explicitly describe a scenario of interaction between this and any of the other 6 short pieces of code in ashift.c that could lead to an unexpected/undesirable outcome and how this is prevented by this lock. I'm really trying to learn here; taking and releasing a lock look like quite expensive operations and unless I really understand this I won't be able to distinguish situations where this effort is wasted superfluously. Forgive me but there's quite a lot of code sprinkled around the codebase where patterns have been copied around cargo-cult-like and refactorings are as a good a time as any to try to get rid of them.

I _think_ I _might_ be getting the point here; the check is only meant to run while nothing else is holding the lock, not to actually protect anything this piece of code might be doing to g->buf (because it doesn't do anything but error checking?) The printfs further down that use g->buf could be printing rubbish, but that's not a big problem because it is only logging code anyway?

Reply:
I was just doing what amounts to search&replace, not analyzing the locking itself.  I'm sure there are places where it could be improved.
And as @parafin pointed out, in many of the places where there's just a single read or write, we'd need at least a memory fence anyway.

Reply:
@dterrahe, I haven't reviewed the whole code of this iop, but here it looks like just a check if g->buf is NULL, so having just one read guarded by mutex seems correct (as you discovered itself, we're not counting ifdefed debug code). But I'm sure there's a lot of incorrect locking happening in dt.

I also want to say, the answer is almost never "lock is not needed here". If you have 2 threads that might run in parallel and access the same data, you must guard those accesses with mutex (or several). If you think that in some particular case it's not needed, you better have a very thought-through (documented in comments) explanation on why not;) Which might still break after becoming obsolete with later code modifications.